### DISCUSSION TITLE
When I User `workspace/symbol` request, response result is empty array

### DISCUSSION THREAD

**Discussion by Yurii-huang** [clangd/clangd#1339] [Q&A]
** Log **
```
I[15:29:39.838] <-- initialize(1)
I[15:29:39.838] --> reply:initialize(1) 0 ms

V[15:29:39.838] >>> {"id":1,"jsonrpc":"2.0","result":{"capabilities":{"astProvider":true,"callHierarchyProvider":true,"codeActionProvider":true,"compilationDatabase":{"automaticReload":true},"completionProvider":{"allCommitCharacters":[" ","\t","(",")","[","]","{","}","<",">",":",";",",","+","-","/","*","%","^","&","#","?",".","=","\"","'","|"],"resolveProvider":false,"triggerCharacters":[".","<",">",":","\"","/"]},"declarationProvider":true,"definitionProvider":true,"documentFormattingProvider":true,"documentHighlightProvider":true,"documentLinkProvider":{"resolveProvider":false},"documentOnTypeFormattingProvider":{"firstTriggerCharacter":"\n","moreTriggerCharacter":[]},"documentRangeFormattingProvider":true,"documentSymbolProvider":true,"executeCommandProvider":{"commands":["clangd.applyFix","clangd.applyTweak"]},"hoverProvider":true,"implementationProvider":true,"memoryUsageProvider":true,"referencesProvider":true,"renameProvider":true,"selectionRangeProvider":true,"semanticTokensProvider":{"full":{"delta":true},"legend":{"tokenModifiers":["declaration","deprecated","deduced","readonly","static","abstract","dependentName","defaultLibrary","functionScope","classScope","fileScope","globalScope"],"tokenTypes":["variable","variable","parameter","function","method","function","property","variable","class","interface","enum","enumMember","type","type","unknown","namespace","typeParameter","concept","type","macro","comment"]},"range":false},"signatureHelpProvider":{"triggerCharacters":["(",","]},"textDocumentSync":{"change":2,"openClose":true,"save":true},"typeHierarchyProvider":true,"workspaceSymbolProvider":true},"serverInfo":{"name":"clangd","version":"clangd version 13.0.0 (https://github.com/llvm/llvm-project d7b669b3a30345cfcdb2fde2af6f48aa4b94845d) linux+grpc x86_64-unknown-linux-gnu"}}}

V[15:29:39.839] <<< {"id":2,"jsonrpc":"2.0","method":"workspace/symbol","params":{"partialResultToken":"{5768ecc8-81b2-4845-a11d-49432b8fb1df}","query":"","workDoneToken":"{22cbcf96-bc87-4bd0-b705-dd903a83462a}"}}

I[15:29:39.839] <-- workspace/symbol(2)
V[15:29:39.839] <<< {"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"languageId":"cpp","text":"#include <iostream>\n\nusing namespace std;\nstruct ABC{\n};\n\nstruct CCC{\n};\n\nint main()\n{\n    std::cout << \"Hello World!\" << std::endl;    \n    ABC *p = new ABC();\n    delete p; delete p;\n    return 0;\n}\n","uri":"file:///home/funning/workspace/testcmake/main.cpp","version":0}}}

I[15:29:39.839] <-- textDocument/didOpen
I[15:29:39.839] --> reply:workspace/symbol(2) 0 ms
V[15:29:39.839] >>> {"id":2,"jsonrpc":"2.0","result":[]}

I[15:29:39.839] Failed to find compilation database for /home/funning/workspace/testcmake/main.cpp
I[15:29:39.839] ASTWorker building file /home/funning/workspace/testcmake/main.cpp version 0 with command clangd fallback
[/home/funning/workspace/testcmake]
/usr/lib/llvm-7/bin/clang -resource-dir=/usr/lib/clang/13.0.0 -- /home/funning/workspace/testcmake/main.cpp
V[15:29:39.839] <<< {"id":3,"jsonrpc":"2.0","method":"textDocument/semanticTokens/full","params":{"textDocument":{"uri":"file:///home/funning/workspace/testcmake/main.cpp"}}}

I[15:29:39.839] <-- textDocument/semanticTokens/full(3)
```

Is it not supported, or is there a precondition request for calling?

Thank anyone

**Comment by HighCommander4** [comment]
This line is relevant:

> I[15:29:39.839] Failed to find compilation database for /home/funning/workspace/testcmake/main.cpp

The project needs to have a compilation database so clangd knows which files to index.

In addition, at least one file needs to be opened first, to trigger clangd to load the index, and after that `workspace/symbol` should work.

**Reply by Yurii-huang** [reply]
Is it incremental? If I open a file, I can only request all symbols in the open file?

> This line is relevant:
> 
> > I[15:29:39.839] Failed to find compilation database for /home/funning/workspace/testcmake/main.cpp
> 
> The project needs to have a compilation database so clangd knows which files to index.
> 
> In addition, at least one file needs to be opened first, to trigger clangd to load the index, and after that `workspace/symbol` should work.

thanks. I have another question
Is it incremental? If I open a file, I can only request workspace symbols in the open file?

**Reply by Yurii-huang** [reply]
about not found compilation database json file:

command line params for:
```
--log=verbose --compile-commands-dir=/home/funning/workspace/testcmake/.unioncode
```
![image](https://user-images.githubusercontent.com/34232854/196374445-7d3bc94f-ed2e-40a7-9426-24be14f18ea1.png)

```
I[15:29:39.839] Failed to find compilation database for /home/funning/workspace/testcmake/main.cpp
```
It should not have this problem

**Reply by Yurii-huang** [reply]
![image](https://user-images.githubusercontent.com/34232854/196376961-40c6168a-4572-416b-9aa9-728c71569f6f.png)

**Reply by HighCommander4** [reply]
@FunningC0217 That does seem strange. Could you share the entire clangd log from startup?

**Reply by HighCommander4** [reply]
@FunningC0217 to answer this:

> Is it incremental? If I open a file, I can only request workspace symbols in the open file?

No, opening the file is just a trigger for clangd to load the index, and then `workspace/symbol` will provide all symbols for the index.

The reason loading the index is tied to opening a file is that the index is stored next to the `compile_commands.json`, and clangd (usually) looks for the `compile_commands.json` based on the path of the open file. (However, this is not the case if `--compile-commands-dir` is used -- I filed https://github.com/clangd/clangd/discussions/1341 about this.)

One could also imagine clangd supporting, as an extension, a protocol message like `clangd/loadIndex`, to load the index at a given path without waiting for a file to be opened.
